File: algorithms/CPlusPlus/Recursion/binary-search.cpp
/*
Description: Binary search using recursion

Approach: To use recursion, for updating start and end variables.

Time Complexity: O(log n)
*/

#include <iostream>
using namespace std;

//function starts
int search(int arr[], int n, int start, int end, int target)
{
    //base case
    //if start is greater than end, it means we haven't found our target element
    //return -1
    if (start > end)
    {
        return -1;
    }
    //calculating mind element
    int mid = start + (end - start) / 2;

    if (arr[mid] == target)
    {
        return mid;
    }
    if (arr[mid] > target)
    {
        //end = mid-1;
        return search(arr, n, start, mid - 1, target);
    }
    else
    {
        //start=mid+1;
        return search(arr, n, mid + 1, end, target);
    }
}

//main starts
int main()
{
    // Write C++ code here
    int arr[] = {1, 2, 3, 4, 5, 6};
    int n = sizeof(arr) / sizeof(arr[0]);
    int target = 6;
    int start = 0;
    int end = n - 1;
    int ind = search(arr, n, start, end, target);
    cout << "Element found at index: " << ind;
    return 0;
}

/*
Sample input:
arr=[1,2,3,4,5,6]
target=6;

Output:
Element found at index: 5
*/

File: algorithms/Java/recursion/binary-search.java
package com.dsa;

/*
Description: Binary Search Program in Java using Recursion

Time Complexity: O(log N)
*/

public class BinarySearch {
    //main starts
    public static void main(String[] args) {
        int arr[] = {12, 19, 23, 45, 50, 67, 89};
        int target = 89;
        int ans = binarySearchR(arr, target, 0, arr.length - 1);
        System.out.println("Element found at index: "+ ans);
    }

    //function
    static int binarySearchR(int[] arr, int target, int start, int end) {
        //finding the mid
        int mid = start + (end - start) / 2;
        //base case, if element not found, return -1
        if (start > end) {
            return -1;
        }
        //return the index if the mid is found
        if (arr[mid] == target) {
            return mid;
        }
        //recursisve function call
        if (arr[mid] > target) {
            return binarySearchR(arr, target, start, mid - 1);
        } else {
            return binarySearchR(arr, target, mid + 1, end);
        }
    }
}
